2164 update config from list of variables to class

lucas-wilkins opened the pull request:
This PR is for a big housekeeping effort surrounding the configuration system.

It does the following things:

1) A new folder called system that replaces the custom_configs/local_config system which contains:
    a) `config` directory, which contains a class called config which controls configuration variables, config_meta describes the behaviour of this class, including loading and saving. More fully documented in the config class pramble.
    b) `web` a class containing all the urls and email addresses
    c) `legal` a class containing any legal stuff, currently just the copyright notice
    d) `env` a utility class for dealing with environment variables
    e) `log` a relocation of `logger_config.py`
  Config settings are now all defined though `config.py` and runtime changes to these variables are written to a local `config.json` file on closing (which is loaded on startup)
   There is now no writing python files, and variable usage is much more easily followed
2) Removal of config items that are no longer used, there were a lot of these.
3) Simplification of the way configuration data is utilised.
 
Things that probably need the most attention when reviewing:
1) Do the GPU settings still integrate properly with your hardware and with sasmodels? (@pkienzle ?)
2) Places issues might occur causing silent errors e.g. in fitting methods where config is used?
3) Have I changed the meaning of anything, or logic of anything that was handled by the local/custom config system?
4) Have I broken anything on other OSs than windows?

Note: this PR addresses multiple issues (linked) which will need to be manually closed.

lucas-wilkins commented on docs/sphinx-docs/source/_extensions/mathjax.py:
```diff
@@ -81,10 +81,10 @@ def html_visit_displaymath(self, node):
     jaxpath = app.config.mathjax_path
     if not jaxpath:
-        raise ExtensionError('mathjax_path config value must be set for the '
+        raise ExtensionError('mathjax_path config_system value must be set for the '
```
urgh, automatic refactoring doesn't always work well

lucas-wilkins commented on docs/sphinx-docs/source/conf.py:
```diff
@@ -41,7 +41,7 @@
 mathjax_path = (
     'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?'
-    'config=TeX-MML-AM_CHTML')
+    'config_system=TeX-MML-AM_CHTML')
```
Definitely needs fixing!!!

lucas-wilkins commented on installers/installer_generator64.py:
```diff
@@ -1,5 +1,5 @@
 """
-This module generates .iss file according to the local config of
+This module generates .iss file according to the local config_system of
```
This file needs deleting anyway

krzywon commented on src/sas/__init__.py:
```diff
@@ -12,10 +16,59 @@ def get_app_dir():
+        return path
+
+    # If we are running frozen, then root is a parent directory
+    if sys.platform == 'darwin':
+        # Here is the path to the file on the mac:
+        #     .../Sasview.app/Contents/Resources/lib/python2.7/site-packages.zip/sas/__init__.pyc
+        # We want the path to the Resources directory.
+        path = dirn(__file__, 5)
+    elif os.name == 'nt':
+        # Here is the path to the file on windows:
+        #     ../Sasview/library.zip/sas/__init__.pyc
+        # We want the path to the Sasview directory.
+        path = dirn(__file__, 3)
```
The depths here are wrong as of v5.0.4 as outlined in #1911. I was looking for a cross-platform solution to this in #2171 using `glob` but wasn't able to 100% solve the issue.

lucas-wilkins replied:
generally, we should be using `importlib.resources` where possible. This is on the list of things to do, but quite a long way down

krzywon commented on src/sas/__init__.py:
```diff
@@ -12,10 +16,59 @@ def get_app_dir():
         _APP_DIR = find_app_dir()
     return _APP_DIR
 
+# TODO: Replace with more idomatic version
+def dirn(path, n):
```
I 100% agree with your TODO here.

krzywon reviewed (commented):
Looking through this, I have a few comments, but the direction you are taking looks good. My related PR (#2167) currently uses `get_custom_config` which you are deprecating, but that looks like the only clash we should have once you're finished.

lucas-wilkins commented on installers/installer_generator.py:
```diff
@@ -11,35 +10,36 @@
-APPLICATION = str(local_config.__appname__ )+ '.exe'
-AppName = str(local_config.__appname__ )
-AppVerName = str(local_config.__appname__ )+'-'+ str(local_config.__version__)
+APPLICATION = str(config.__appname__) + '.exe'
+AppName = str(config.__appname__)
+AppVerName = str(config.__appname__) + '-' + str(config.__version__)
```
At the moment this is all just copied straight from ```config_*.py```, will be making substantial changes to it

lucas-wilkins commented on installers/installer_generator.py:
```diff
@@ -11,35 +10,36 @@
-icon_path =  local_config.icon_path
-media_path = local_config.media_path
-test_path = local_config.test_path
+icon_path =  config.icon_path
+media_path = config.media_path
+test_path = config.test_path
```
yup, config shouldn't have any paths in it anyway.

lucas-wilkins commented on installers/installer_generator64.py:
```diff
@@ -1,7 +1,6 @@
 """
 This module generates .iss file according to the local config of
```
yeah, its going eventually, as is the non-64 one, as @rozyczko informs me that its not used either.

lucas-wilkins commented:
Status update:

I've finished making the config class, the next task is look the variables within it, throw out unused ones, replace bad uses of a config with more proper usages etc. Basically I've done steps 1, 2, 5 and 7 from my initial list.

Appropriate loading and saving on startup is yet to be implemented, basically because I'm not sure about where the user config file (which is now the only config file that should exist) should live.

There is pretty decent test coverage for these changes.

@krzywon

lucas-wilkins commented:
@krzywon Here's the preamble from the config.py which explains the reasoning behind it


Configs
=======

Configs are a nightmare from the perspective of code maintainability. There are 
three main reasons for this
  1) They have a tendency to accumulate junk because people don't realise that a
     config item is no longer needed
  2) It's hard to trace the usages and types because values are loaded at runtime
  3) Maintaining synchrony between config files and config usages is difficult, as
     it is the users that have control over the config files.

The Config class here attempts to resolve some of these issues in a way that
preserves as many of the current uses as possible. It is a compromise between
SasView's current methods, and more standard ways of handling configurations.

Brief Outline
=============

The main Config class provides a definition of the variables allowed in a config
file, along with their default values. This class is used to generate a schema
that determines how config files are read. Only a few types of variable are allowed

* bool
* int
* float
* str
* Homogeneous lists of the above, to 10 levels of depth
* None and empty list (please try to avoid)

Other types will throw an error at runtime when the schema is created.

None types and
empty lists have no type information at runtime, so the schema cannot check/coerce
the type of config variables when they are loaded. It is best to avoid having these
as default values if possible.

The presence of a config file is not necessary for the functioning of the config
system, only for making changes that differ from the default values.


What Belongs in a Config
========================

Things that do belong:
  1) Program settings that are configurable by users through the GUI
  2) Program settings that have no GUI editor, but that some advanced users
     might want to set manually with a text editor
  3) Settings that control developer tools, e.g. debug modes
  4) Very little else

Things that don't belong, but were previously in the config:
  1) dynamic content, i.e. values that are modified programmatically,
    this includes variables that are defined in terms of other variables,
    but otherwise don't change
  2) Paths to resources within sasview (use importlib.resources instead)
  3) Blocks of data that won't be modified by the user and used primarily
     by single class - e.g. the text for a message
  4) Large blocks of text in general

Making Changes to the Config Class
==================================

As users have their own copy of the sasview configuration, deletions,
name changes, default value changes, and variable type changes
often constitute a breaking change from the perspective of version
control. The users locally stored config will, in general, not be
backwards compatible with the new config. Extreme caution should be
exercised - when changing the config, don't just think about the
problem at hand, but about the future maintainability of SasView in
general.

Deleting from Config class:
Currently (02/09/2022) the consequences of providing an entry in a
config file that is not properly reflected in the Config class is a
error. To ease backward compatibility, it is possible to disable
the errors for a deleted class member by adding their name to
the `_deleted_attributes` list. The suggested deletion process would
be

```
[-]   self.my_variable = 10
[+]   self._deleted_attributes.append("my_variable")
```

At major releases, additions to _deleted_attributes should be removed.


Other Design Decisions made for the Config Class
================================================

The Config class cannot be dynamically modified, this prevents the config
from having fields that are unspecified in the base class,
and makes it so that all usages of fields can be automatically tracked.

Subclassing of Config is also disabled for similar reasons.

I have opted not to use frozen dataclasses at this time
because, as they currently work, the behaviour would make creating
different configs difficult.

lucas-wilkins commented:
See https://github.com/SasView/sasview/issues/2176 for why your tests are failing, merge the associated branch for fix, or merge main when https://github.com/SasView/sasview/pull/2178 is merged in.

Will merge this in when I'm back at work

lucas-wilkins commented:
@krzywon This is ready.

krzywon commented on src/sas/system/web.py:
```diff
@@ -0,0 +1,24 @@
+class WebLinks:
+    def __init__(self):
+        self.nist_url = "https://www.nist.gov/"
```
`https://www.nist.gov/ncnr`

lucas-wilkins replied:
Done

krzywon commented on src/sas/system/web.py:
```diff
@@ -0,0 +1,24 @@
+        self.diamond_url = "http://www.diamond.ac.uk"
+
+        self.homepage_url = "https://www.sasview.org"
+        self.download_url = 'https://github.com/SasView/sasview/releases'
```
This brings users to a list of all releases, including the test builds.

`'https://github.com/SasView/sasview/releases/latest'` would probably be better.

lucas-wilkins replied:
I just straight up copied these, but happy to update now.

lucas-wilkins replied:
Done

krzywon commented on src/sas/system/web.py:
```diff
@@ -0,0 +1,24 @@
+
+        self.homepage_url = "https://www.sasview.org"
+        self.download_url = 'https://github.com/SasView/sasview/releases'
+        self.marketplace_url = "http://marketplace.sasview.org/"
```
If we can sort out sasview/sasmodel-marketplace#29 soon, this should become `https://marketplace.sasview.org/`

lucas-wilkins replied:
ok, well you know where to change it when the time comes.

krzywon commented on src/sas/system/config/schema_elements.py:
```diff
@@ -0,0 +1,236 @@
+        return SchemaNonSpecified()
+
+    # List case
+    elif isinstance(value, list):
```
SchemaList coerces tuples and sets to a list (with a warning), but your creation method only accepts lists. Any reason for this?

lucas-wilkins replied:
If I understand you correctly, you're asking why when parsing the fields in `Config` it only accepts lists, but one can coerce a bunch of different collections. This behaviour allows someone to set a config variable which was a list, using a tuple, for instance, even though the default value was a list. I think its fine, but also being stricter would be fine too.

krzywon replied:
I think I understand. If someone sets a value in config.py to a tuple or set, it will be coerced to a list (with a warning), so only lists are stored in and loaded from config.json, thus only allowing lists here.

krzywon reviewed (changes requested):
I tested most of the functionality and everything seems to be working, including storing new values as they are changed. I didn't look through the unit tests as closely as I could have, but they are there and they seemingly pass. Overall, this looks pretty good, but I do have a couple of comments and concerns.

- The version issue I noted should be fixed.
- The commented code should probably be removed.
- Someone should review the unit tests.
- URL suggestions might management input (@butlerpd, @smk78)
- Other suggestions I leave to you.

lucas-wilkins commented on src/sas/__init__.py:
```diff
@@ -1,5 +1,14 @@
+
 
 _APP_DIR = None
 def get_app_dir():
```
Removed and more!

lucas-wilkins commented on src/sas/__init__.py:
```diff
@@ -1,5 +1,14 @@
-__all__ = ['get_app_dir', 'get_user_dir',
-           'get_local_config', 'get_custom_config']
+from sas.sasview import __version__
```
I've changed quite a bit here, basically added some more files to system for all this. Hopefully the circular import is resolved now?

lucas-wilkins commented on setup.py:
```diff
@@ -42,24 +42,24 @@
-    #         if f in plugin_model_list:
-    #             file_path =  os.path.join(f_path, f)
-    #             os.remove(file_path)
+# # TODO: build step should not be messing with existing installation!!
```
Going to have a proper look at setup soon, leaving for now.

krzywon commented on src/sas/qtgui/MainWindow/GuiManager.py:
```diff
@@ -176,7 +178,7 @@ def addWidgets(self):
 
-        self._workspace.toolBar.setVisible(LocalConfig.TOOLBAR_SHOW)
+        self._workspace.toolBar.setVisible(config.TOOLBAR_SHOW)
         self._workspace.actionHide_Toolbar.setText("Show Toolbar")
```
The toolbar is shown by default, but the test is defaulting to the hidden text. This line should get removed.

lucas-wilkins replied:
Don't get what you're saying here.

krzywon replied:
`.setVisible()` sets the visibility and the menu text for the toolbar. Line 182, right after `setVisible()` is called, sets the menu text to the text shown when the toolbar is hidden, regardless of the toolbar visibility. Line 182 should probably be deleted to eliminate the potential inconsistency when the toolbar is shown by default.

krzywon commented on src/sas/system/config/config.py:
```diff
@@ -0,0 +1,196 @@
+        self.SAS_OPENCL = "none"
+
+        self.DEFAULT_OPEN_FOLDER = ""
+        self.TOOLBAR_SHOW = True
```
On second glance, many of these booleans aren't saved in config.json and probably should. Here is an initial list that I will update as I find more and as they are fixed.

- TOOLBAR_SHOW
- VIEW_MENU
- whatever is used for showing/hiding the console log
- SHOW_WELCOME_PANEL is set to False in the config, but the welcome panel is still shown

lucas-wilkins replied:
It won't get saved in config.json unless its been saved as something other than the default value.

lucas-wilkins replied:
The welcome panel isn't what you think it is

krzywon replied:
I hid the toolbar, console log, and data explorer, all of which are shown by default. I closed the program, but none of those changes were stored in config.json, so they were all visible when I opened the program again.

krzywon replied:
> The welcome panel isn't what you think it is

You're right. I mixed up the splash screen and welcome panel.

lucas-wilkins replied:
I will make a new issue for these

lucas-wilkins replied:
#2208

lucas-wilkins commented on src/sas/qtgui/Utilities/CustomDir.py:
```diff
@@ -1,58 +1,56 @@
-    return os.path.join(u_dir, CONF_DIR)
+    return os.path.join(u_dir, config.CONF_DIR)
 
 def setup_conf_dir(path):
```
Certainly

lucas-wilkins commented on src/sas/system/config/config_meta.py:
```diff
@@ -0,0 +1,182 @@
+            raise MalformedFile("Malformed config file - no 'sasview_version' key")
+
+        try:
+            parts = [int(s) for s in data["sasview_version"].split(".")]
```
`StrictVersion` is run elsewhere - now in version.py

lucas-wilkins commented:
Ah, I see what your referring to. Pretty sure none of them ever worked like
that. Would be pretty simple to implement now though.

On Thu, 15 Sept 2022, 15:47 Jeff Krzywon, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/sas/system/config/config.py
> <https://github.com/SasView/sasview/pull/2168#discussion_r972084778>:
>
> > +
> +        self.SHOW_WELCOME_PANEL = False
> +        # OPEN and SAVE project menu
> +        self.OPEN_SAVE_PROJECT_MENU = True
> +        # VIEW MENU
> +        self.VIEW_MENU = True
> +        # EDIT MENU
> +        self.EDIT_MENU = True
> +
> +
> +
> +        # OpenCL option - should be a string, either, "none", a number, or pair of form "A:B"
> +        self.SAS_OPENCL = "none"
> +
> +        self.DEFAULT_OPEN_FOLDER = ""
> +        self.TOOLBAR_SHOW = True
>
> I hid the toolbar, console log, and data explorer, all of which are shown
> by default. I closed the program, but none of those changes were stored in
> config.json, so they were all visible when I opened the program again.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SasView/sasview/pull/2168#discussion_r972084778>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACKU4SRVJOZ4C3R5JKHVKGLV6MZIRANCNFSM57CLHS2Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

lucas-wilkins commented:
Ok, I see. Can fix that.

On Thu, 15 Sept 2022, 15:42 Jeff Krzywon, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/sas/qtgui/MainWindow/GuiManager.py
> <https://github.com/SasView/sasview/pull/2168#discussion_r972078275>:
>
> > @@ -176,7 +178,7 @@ def addWidgets(self):
>          self.results_frame.setVisible(False)
>          self.results_panel.windowClosedSignal.connect(lambda: self.results_frame.setVisible(False))
>
> -        self._workspace.toolBar.setVisible(LocalConfig.TOOLBAR_SHOW)
> +        self._workspace.toolBar.setVisible(config.TOOLBAR_SHOW)
>          self._workspace.actionHide_Toolbar.setText("Show Toolbar")
>
> .setVisible() sets the visibility and the menu text for the toolbar. Line
> 182, right after setVisible() is called, sets the menu text to the text
> shown when the toolbar is hidden, regardless of the toolbar visibility.
> Line 182 should probably be deleted to eliminate the potential
> inconsistency when the toolbar is shown by default.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SasView/sasview/pull/2168#discussion_r972078275>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACKU4SXCXLN6OU2IYDLNM5TV6MYW5ANCNFSM57CLHS2Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

lucas-wilkins commented on src/sas/qtgui/Utilities/GuiUtils.py:
```diff
@@ -68,24 +70,18 @@
-    if os.path.isfile(os.path.join(os.getcwd(), "custom_config.py")):
-        #logging.info("Using application path: %s", os.getcwd())
-        return os.path.abspath(os.getcwd())
+    #
```
Done

krzywon commented on src/sas/qtgui/MainWindow/GuiManager.py:
```diff
@@ -570,16 +573,15 @@ def processVersion(self, version_info):
                 self.communicate.statusBarUpdateSignal.emit(msg)
 
-            elif version.__gt__(LocalConfig.__version__):
+            elif version.__gt__(sas.system.version.__version__):
```
`sas.system.version.__version__` is imported as `SASVIEW_VERSION`. Any reason to not use the constant?

lucas-wilkins replied:
Shouldn't really be imported like that. I saw it, but I have kept it in so as to avoid mission creep.

krzywon commented on src/sas/qtgui/MainWindow/PackageGatherer.py:
```diff
@@ -146,7 +147,7 @@ def get_imported_packages(self):
                 # Module has __version__ attribute
                 try:
-                    package_versions_dict[package_name] = package.__version__
+                    package_versions_dict[package_name] = sas.system.version.__version__
```
This change should be reverted. This was storing the version of each dependency.

lucas-wilkins replied:
yeah, didn't spot that.

krzywon commented on src/sas/system/web.py:
```diff
@@ -0,0 +1,24 @@
+        self.ill_url = "https://www.ill.eu/"
+        self.ansto_url = "https://www.ansto.gov.au/"
+        self.bam_url = "http://www.bam.de/"
+        self.delft_url = "http://www.tudelft.nl/en/tnw/business/facilities/reactor-instituut-delft/"
```
As noted by @smk78 in #2200:

> All the links from the About box work, but I notice that TU Delft have made a change so that URL should now redirect to:
https://www.tudelft.nl/en/faculty-of-applied-sciences/business/facilities/reactor-institute-delft

krzywon commented on src/sas/qtgui/Utilities/UnitTesting/GuiUtilsTest.py:
```diff
@@ -46,12 +46,6 @@ def testGetAppDir(self):
         """
         pass
 
-    def testGetUserDirectory(self):
```
Suggestion: Move this test into your config tests.

lucas-wilkins replied:
Maybe if the tests are refactored more generally so that everything in system is in the same place.

lucas-wilkins replied:
oh, actually, this is part of the gui testing suite, which is currently a very different place to the config tests.

krzywon reviewed (commented):
I'm still seeing a few issues. I've also now reviewed your unit tests and only have a single suggestion on them (see my individual comments).

Running `python run.py` after running `python setup.py clean build install`, I get an error that images in the sas/qtgui/images directory cannot be found. I haven't had a chance to trace this, but I don't see any time that directory was included with the build. This error is not present when running the `main` branch, though, so I'm not sure why it's happening now.

> Traceback (most recent call last):
>   File "\~\devel\sasview\run.py", line 103, in <module>
>     import sas.qtgui.convertUI
>   File "\~\.conda\envs\qt5_win_py39\lib\site-packages\sasview-5.0.5-py3.9.egg\sas\qtgui\convertUI.py", line 83, in <module>
>     if file_in_newer(in_file, out_file):
>   File "\~\.conda\envs\qt5_win_py39\lib\site-packages\sasview-5.0.5-py3.9.egg\sas\qtgui\convertUI.py", line 46, in file_in_newer
>     in_stat = os.stat(file_in)
> FileNotFoundError: [WinError 3] The system cannot find the path specified: '\~\\.conda\\envs\\qt5_win_py39\\lib\\site-packages\\sasview-5.0.5-py3.9.egg\\sas\\qtgui\\images\\images.qrc'

Running the executable generated by the github workflow, the following error pops up and SasView fails to start. This might be a `setuptools` version issue and might not be exclusive to your branch, but I haven't tested other recently built executables.

> Traceback (most recent call last):
>   File "sasview.py", line 13, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
>   File "sas\__init__.py", line 1, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
>   File "sas\system\__init__.py", line 4, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
>   File "sas\system\config\config.py", line 133, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
>   File "sas\system\config\config_meta.py", line 8, in <module>
>   File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
>   File "sas\system\version.py", line 1, in <module>
>   File "_distutils_hack\__init__.py", line 88, in create_module
>   File "importlib\__init__.py", line 127, in import_module
>   File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
>   File "setuptools\__init__.py", line 10, in <module>
>   File "_distutils_hack\__init__.py", line 88, in create_module
>   File "importlib\__init__.py", line 127, in import_module
> ModuleNotFoundError: No module named 'setuptools._distutils'

lucas-wilkins commented on run.py:
```diff
@@ -122,13 +96,8 @@ def prepare():
     root = abspath(dirname(realpath(sys.argv[0])))
 
     addpath(joinpath(root, 'src'))
-    addpath(joinpath(root, joinpath('..', 'sasmodels'))) # dependency (for loading custom_config.py during log setup)
```
I assume from the comment it doesn't matter.

lucas-wilkins commented on run.py:
```diff
@@ -73,15 +72,6 @@ def prepare():
-    # matplotlib.use('Agg')
-    # print matplotlib.__file__
-    #import pylab; pylab.hold(False)
-    # add periodictable to the path
```
It was somewhat deliberate

lucas-wilkins replied:
> Running python run.py after running python setup.py clean build install, I get an error that images in the sas/qtgui/images directory cannot be found. I haven't had a chance to trace this, but I don't see any time that directory was included with the build. This error is not present when running the main branch, though, so I'm not sure why it's happening now.

I think this is an issue with `setup.py`, this kind of thing is going to keep happening until its been shown some love. For now, it can probably just be added. Fixing the issues with `setup.py` is next on my list after this PR is accepted.

> Running the executable generated by the github workflow, the following error pops up and SasView fails to start. This might be a setuptools version issue and might not be exclusive to your branch, but I haven't tested other recently built executables.

Ah yeah, that's because of StrictVersion - I don't think it's really needed if we do sensible versioning. Should be an easy fix.

lucas-wilkins replied:
Not really accidental.

lucas-wilkins commented:
> Running python run.py after running python setup.py clean build install, I get an error that images in the sas/qtgui/images directory cannot be found. I haven't had a chance to trace this, but I don't see any time that directory was included with the build. This error is not present when running the main branch, though, so I'm not sure why it's happening now.

I think this is an issue with `setup.py`, this kind of thing is going to keep happening until its been shown some love. For now, it can probably just be added. Fixing the issues with `setup.py` is next on my list after this PR is accepted.

> Running the executable generated by the github workflow, the following error pops up and SasView fails to start. This might be a setuptools version issue and might not be exclusive to your branch, but I haven't tested other recently built executables.

Ah yeah, that's because of StrictVersion - I don't think it's really needed if we do sensible versioning. Should be an easy fix.

krzywon commented on run.py:
```diff
@@ -20,6 +20,8 @@
 from os.path import abspath, dirname, realpath, join as joinpath
 from contextlib import contextmanager
 
+import sas.system.user
```
I removed this import locally (it isn't used) and `run.py` successfully started SasView.

>> Running python run.py after running python setup.py clean build install, I get an error that images in the sas/qtgui/images directory cannot be found. I haven't had a chance to trace this, but I don't see any time that directory was included with the build. This error is not present when running the main branch, though, so I'm not sure why it's happening now.

>I think this is an issue with setup.py, this kind of thing is going to keep happening until its been shown some love. For now, it can probably just be added. Fixing the issues with setup.py is next on my list after this PR is accepted.

krzywon reviewed (approved):
The issue with running the installed version of the package is still present, but the issue isn't limited to this branch. I will put in a new issue describing what I think is going on. Run.py works again.

I think this is ready, but want at least one more set of eyes to look over this before merging.

lucas-wilkins commented:
> I think this is ready, but want at least one more set of eyes to look over this before merging.

@pkienzle looked it over, it just doesn't show as I clicked the re-request button. There were not any big changes to make from him.

pkienzle reviewed (commented):
Mostly suggestions that you can ignore.

You should sort out the circular reference between setup logger and config.load, preferably by moving config.load after setup logger and out of the root sas namespace.

There are still some imports of sas.sasview that can be removed.

lucas-wilkins commented on build_tools/release_automation.py:
```diff
@@ -135,29 +135,31 @@ def generate_zenodo(sasview_data, zenodo_api_key):
+__version__ = "%s"
+__release_date__ = "%i"
+__build__ = "GIT_COMMIT"
+"""
```
Didn't know you could do that. I agree its better.

lucas-wilkins commented on src/sas/qtgui/MainWindow/AboutBox.py:
```diff
@@ -26,20 +29,38 @@ def addText(self):
+                <p>
+                    Comments? Bugs? Requests?
+                    <br/>
+                    <a href="mailto:{web._license}">Send us a ticket</a>
```
It's what it was called before, I meant to rename it.

lucas-wilkins commented on src/sas/qtgui/MainWindow/MainWindow.py:
```diff
@@ -1,9 +1,12 @@
-from sas.sasview import __version__ as SASVIEW_VERSION
+from ...system.version import __version__ as SASVIEW_VERSION
+from sas import config
+from sas.system import env
```
automatic refactoring.

lucas-wilkins commented on src/sas/system/log.py:
```diff
@@ -46,17 +46,7 @@ def config_development(self):
-
-        if disable_debug:
-            # logging.info("Note: DEBUG logs are disabled.")
+        if config.FILTER_DEBUG_LOGS:
```
Take it this was the circular reference you were referring to.

I'm using logging in config to give a warning message, one that should probably be fixed by the programmer. Will look into alternatives. I might just use warnings.

lucas-wilkins replied:
I've reviewed it, I think its fine.

pkienzle replied:
`config_meta.load_from_file_object` (which should probably be `_load_from_file_object`) can call `logger.warning`. With `config.load()` called in `sas.__init__` this happens before `SetupLogger` is called from `MainWindow.run_sasview`.

You can replace the `logger.warning` with `warnings.warn` but that doesn't do any good since I suspect the windows GUI startup eats everything written to stdout without showing it to the user (and I'm guessing many mac users don't know that this info is available through console.app; no idea what the various linux window managers do with it). You do get the warning on save since you have suppressed writing the config file on major version change* so you can just drop the warning here. Or maybe issue a warning after the logger is set up. Note that you should also log the path to the config.json that was loaded, or a list of paths that was searched when it isn't found. It'll make it easier to debug when users complain that their configuration changes were not being saved because we can't find it on load.

I still recommend that you remove the reference from config within the logger so that the order of `config.load` and `SetupLogger` doesn't matter.

Regarding the `config.load` in `sas.__init__`, I have a strong preference for not doing any work when importing, especially for high level imports. I would not expect `import sas` to create a `~/.sasview` directory or a `~/SasView.log` file for example. It looks like the current code does not do so, but it has in the past and I can imagine something like that slipping in later. Also, if you restrict the import to parsing and trivial initialization you don't have to deal with unrelated exceptions on import.

* The "disabled write" warning should tell users to delete the old config file to make the message go away, with full path so they can figure out where cwd is pointing. Another problem with config.json being written to cwd is that this might be some system directory where the user has no write permission. If the current working directory changes when running the program (through a Qt file dialog maybe?) then the saved config.json will not be in the location where load config.json is looking for it.

lucas-wilkins replied:
>I still recommend that you remove the reference from config within the logger so that the order of config.load and SetupLogger doesn't matter.

I'm happy with the logger not doing all the weird config stuff that it currently does, that's a legacy thing that I preserved. In my opinion, if someone wants to change the logging behaviour it should really be done through the `.ini`. I just considered this change beyond the scope of this PR. I'll change it.

>  Another problem with config.json being written to cwd is that this might be some system directory where the user has no write permission. If the current working directory changes when running the program (through a Qt file dialog maybe?) then the saved config.json will not be in the location where load config.json is looking for it.

I'd just written it locally for the time being, will move it to `.sasview/config-<major version>.json`

pkienzle replied:
Okay, I see that this is a pre-existing issue. I was only looking at the code you added, not the code that you deleted.

Given that there is no way to adjust the `FILTER_DEBUG_LOGS` flag in the GUI or command line and you have said that the user is not supposed to edit `config.json` I think it is safe to disable reference to the config parameter and do the default action `logging.disable(logging.DEBUG)`.

lucas-wilkins replied:
Done.

lucas-wilkins commented on src/sas/system/log.py:
```diff
@@ -83,8 +73,8 @@ def _find_config_file(self, filename="logging.ini"):
-            os.path.join("sas", "sasview", filename),
-            os.path.join(os.getcwd(), "sas", "sasview", filename),
+            os.path.join("sas", "system", filename),
+            os.path.join(os.getcwd(), "sas", "system", filename),
```
yeah, I just modified the existing paths, I considered anything beyond that out of scope.

lucas-wilkins commented on src/sas/system/config/config_meta.py:
```diff
@@ -0,0 +1,183 @@
+        if int(file_major_version) != int(sasview_major_version):
+            logger.warning(f"Attempting to used outdated config file (config is"
+                           f" for {file_version}, this SasView version is {sas.system.version.__version__})")
+            self._write_disabled = True
```
It **should** save the config in the startup path, so different versions wont share the same config.

> can you open an newer version of the config file with an older version of sasview

No. It is backward compatible, not forward compatible, as you say. It really shouldn't be forward compatible. The warning is just there for developers who will likely change the major version, but keep the location of the code the same.

The point is *exactly* that old keys shouldn't be hanging around. I spent two weeks finding all the old config variables that were no longer use and were just hanging around, messing up the code. The whole point of this PR is to minimise the accumulation of junk config variables, and make it easier to delete them.

pkienzle replied:
The use case for old and new versions of sasview coexisting in the machine is real. Scientists are distrustful sorts, and don't want to commit to the new without comparing to the old.

I was assuming that `config.json` was loaded/saved from somewhere in `~/.sasview`. Looking at the code in `config_meta.py` it appears to be loaded/saved from the current working directory?!?

I don't see any protection against version collision except as an accident of the O/S startup context. This seems error-prone. And I don't think we want protection from version collision. If I'm dutifully keeping sasview up to date I don't want to reset my configuration each time there is a new version.

Furthermore, I'm specifically suggesting that you do want forward and backward compatibility. You can do this by retaining all the keys in the config file for round-tripping, ignoring new keys in the old version and ignoring old keys in the new version. I'm not suggesting that you maintain access properties for these keys in the config structure. The remainder of the code will only see the available keys for that version. Missing keys (i.e., loading an old config in a new version) should be filled with default values, and saved to the config file on exit. These will be ignored when loading the config from the older version.

lucas-wilkins replied:
> I was assuming that config.json was loaded/saved from somewhere in ~/.sasview. Looking at the code in config_meta.py it appears to be loaded/saved from the current working directory?!?

See my comment above

> Furthermore, I'm specifically suggesting that you do want forward and backward compatibility. You can do this by retaining all the keys in the config file for round-tripping, ignoring new keys in the old version and ignoring old keys in the new version. I'm not suggesting that you maintain access properties for these keys in the config structure. The remainder of the code will only see the available keys for that version. Missing keys (i.e., loading an old config in a new version) should be filled with default values, and saved to the config file on exit. These will be ignored when loading the config from the older version.

The problem with that, one which I was explicitly trying to avoid, is that it won't track changes to default values correctly. If you change a default value with the system you suggest, everyone has to update their config file to see the default behaviour. With the system as I have implemented it, if you have the default behaviour selected and that default changes, your version will reflect that. It's a choice with pros and cons, but my experience is that keeping every developers version as in-sync as possible makes things much easier in the long run.

Another the problem with keeping keys around is that it only takes someone to ignore a variable like `SHOW_THINGY` and make something very similar, `SHOW_THING` to do the same or a similar thing. Now there's two with almost the same name, nobody can remember which one was which, people complain that setting the variable does nothing because they've just changed the first one they see, and nobody wants to fix it because its loads of effort to find out what each one does. This is the motivation behind having some kind of versioning - it's a built in opportunity to fix this sort of thing. Whether or not it needs to complain so hard about the wrong version of the file is something that remains to be tested in practice, but as I mentioned before, it probably makes sense to have a different config file for each major version anyway.

In this PR, specifying an invalid key doesn't break anything, it just gives you log message. It will, however, forget it - I can see how that could potentially be a problem in the future. I will have it retain unknown keys and write them out afterwards.

pkienzle replied:
>  it won't track changes to default values correctly.

This is why I specified that "you should never change how they are used" 8-)

To support changing defaults your can capture the stored keys on load and tweak the save so that the value is stored if it was there before or if it is different from the default:
```python
def load_from_file_object(self, file):
    data = json.load(file)
    self._stored_keys = set(data)
    ...

def save_to_file_object(self):
    data = {
        key: value for key in self._defaults for value in [getattr(self, key)]
        if key in self._stored_keys or value != self._defaults[key]
    }
    ...
```

lucas-wilkins replied:
That won't quite work right, but I will do something similar.

pkienzle replied:
Oops… this doesn't preserve unused keys. Try:
```python
def load_from_file_object(self, file):
    data = json.load(file)
    self._stored = data
    ...

def save_to_file_object(self):
    changed = {key: value for key, default in self._defaults.items() if default != (value := getattr(self, key))}
    data = {**self._stored, **changed}
    ...
```

This is different from the behaviour of your algorithm when using a single version. With your code if you turn on OpenCL then it will save the "use opencl" flags, but if you turn it off again it will match the default and save nothing. This new algorithm will preserve the "don't use opencl" flags when you turn it off. This is only the case if you turn it off in a different session than you turn it on. If it is the same session then the "on" values won't be in stored, so turning it off again won't trigger stored test or the non-default test.

A quick check shows that the key order in stored is preserved in the dict union constructor, though I'm not sure it matters since we are not round-tripping with preserved formatting and comments. And since we are not preserving formatting, we may want to sort keys as well as pretty-print on dump for debugging convenience.

lucas-wilkins commented on src/sas/system/config/config_meta.py:
```diff
@@ -0,0 +1,183 @@
+            "sasview_version": sas.system.version.__version__,
+            "config_data": data}
+
+        json.dump(output_data, file)
```
> Do we want to pretty print on dump so they are easier for the user to read?

Not really for anyone to look at, though might be useful for debugging.

> Are these files ever intended to be modified by the user?

No

> If so, you probably want to allow comments but JSON doesn't allow them. Probably too late to consider an ini format or a yaml format.

I did consider yml, and json is compatible with it. However, it's not meant to be edited manually.

lucas-wilkins commented on src/sas/qtgui/Utilities/FileConverter.py:
```diff
@@ -23,7 +23,7 @@
 from sas.qtgui.Utilities.FrameSelect import FrameSelect
 
-from sas.sasview import __version__ as SASVIEW_VERSION
+from ...system.version import __version__ as SASVIEW_VERSION
```
sure

lucas-wilkins commented on build_tools/release_automation.py:
```diff
@@ -135,29 +135,31 @@ def generate_zenodo(sasview_data, zenodo_api_key):
-    output_lines = []
+    system_directory = "sasview/src/sas/system"
+    version_filename = os.path.join(system_directory, "version.py")
+    zenodo_filename = os.path.join(system_directory, "zenodo.py")
```
.yeah, I think `release_automation.py` could be better in general. My only concern in this PR is keeping it functioning.

lucas-wilkins commented:
@pkienzle @krzywon Will merge this tomorrow if there are no further changes.

lucas-wilkins commented:
Will probably enable pretty printing when I have it open.

On Wed, Sep 28, 2022 at 10:16 PM Paul Kienzle ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/sas/system/config/config_meta.py
> <https://github.com/SasView/sasview/pull/2168#discussion_r982869007>:
>
> > +        if "config_data" not in data:
>
> +            raise MalformedFile("Malformed config file - no 'config_data' key")
>
> +
>
> +        if not isinstance(data["config_data"], dict):
>
> +            raise MalformedFile("Malformed config file - expected 'config_data' to be a dictionary")
>
> +
>
> +
>
> +        # Check major version
>
> +        file_version = data["sasview_version"]
>
> +        file_major_version = file_version.split(".")[0]
>
> +        sasview_major_version = sas.system.version.__version__.split(".")[0]
>
> +
>
> +        if int(file_major_version) != int(sasview_major_version):
>
> +            logger.warning(f"Attempting to used outdated config file (config is"
>
> +                           f" for {file_version}, this SasView version is {sas.system.version.__version__})")
>
> +            self._write_disabled = True
>
>
> Oops… this doesn't preserve unused keys. Try:
>
> def load_from_file_object(self, file):
>
>     data = json.load(file)
>
>     self._stored = data
>
>     ...
>
>
> def save_to_file_object(self):
>
>     changed = {key: value for key, default in self._defaults.items() if default != (value := getattr(self, key))}
>
>     data = {**self._stored, **changed}
>
>     ...
>
> This is different from the behaviour of your algorithm when using a single
> version. With your code if you turn on OpenCL then it will save the "use
> opencl" flags, but if you turn it off again it will match the default and
> save nothing. This new algorithm will preserve the "don't use opencl" flags
> when you turn it off. This is only the case if you turn it off in a
> different session than you turn it on. If it is the same session then the
> "on" values won't be in stored, so turning it off again won't trigger
> stored test or the non-default test.
>
> A quick check shows that the key order in stored is preserved in the dict
> union constructor, though I'm not sure it matters since we are not
> round-tripping with preserved formatting and comments. And since we are not
> preserving formatting, we may want to sort keys as well as pretty-print on
> dump for debugging convenience.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SasView/sasview/pull/2168#discussion_r982869007>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACKU4SRHKLQRDLNQRNWINTDWASYR3ANCNFSM57CLHS2Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 

Dr Lucas Wilkins
+44 (0) 7505 915 726

Personal Website: http://www.lucaswilkins.com/
Alternate e-mail: ***@***.***

lucas-wilkins commented:
@krzywon @krzywon Right, I think its ready!

pkienzle commented on src/sas/system/config/config_meta.py:
```diff
@@ -42,9 +43,16 @@ def __init__(self):
         self._defaults: Dict[str, SchemaElement] = {}
         self._deleted_attributes: List[str] = []
-        self._write_disabled = False
+        self._bad_entries: Dict[str, Any] = {}
```
A tad pejorative: Just because it is unused in the current version doesn't mean it is "bad" 8-)  I suggest a more neutral `_unused_entries`. Also, if you allow unused entries to be propagated I'm not sure you need to handle deleted attributes specially.

lucas-wilkins replied:
deliberately pejorative.

lucas-wilkins replied:
Although there is an argument for allowing different versions to be installed, people shouldn't have a need to use an older version unless we've messed up (I agree this is quite possible). They are bad keys, created from badness.

lucas-wilkins commented on src/sas/qtgui/MainWindow/MainWindow.py:
```diff
@@ -83,7 +83,8 @@ def run_sasview():
         logging.getLogger(__name__).info("Getting OpenCL settings from environment variables")
-        config.SAS_OPENCL = env.sas_opencl
+        if env.SAS_OPENCL != "none":
+            config.SAS_OPENCL = env.sas_opencl
```
Yeah, forgot to update MainWindow

lucas-wilkins commented on src/sas/system/config/config_meta.py:
```diff
@@ -74,20 +83,17 @@ def update(self, data: Dict[str, Any]):
         """ Save config file
 
-        Only changed variables will be included in the saved file
+        Only changed and unknown variables will be included in the saved file
```
Yes. That will happen. I think its worth it to not have "sticky" config entries. I know its not how you would like to see it done, and maybe it will turn out to be a bad way, but I've seen enough problems doing it the way you suggest to think its worth trying to do it differently.
